<!DOCTYPE html>
<html>
<head>
    <title> Async And Await </title>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css" integrity="sha512-DTOQO9RWCH3ppGqcWaEA1BIZOC6xxalwEsw9c2QQeAIftl+Vegovlnee1c9QX4TctnWMn13TZye+giMm8e2LwA==" crossorigin="anonymous" referrerpolicy="no-referrer" />
    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-rbsA2VBKQhggwzxH7pPCaAqO46MgnOM80zW1RWuH61DGLwZJEdK2Kadq2F9CUG65" crossorigin="anonymous">

    <style type="text/css">

    </style>
</head>
<body>

<script src="https://cdn.jsdelivr.net/npm/@popperjs/core@2.11.6/dist/umd/popper.min.js" integrity="sha384-oBqDVmMz9ATKxIep9tiCxS/Z9fNfEXiDAYTujMAeBAsjFuCZSmKbSSUnQlmh/jp3" crossorigin="anonymous"></script>
<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/js/bootstrap.min.js" integrity="sha384-cuYeSxntonz0PPNlHhBs68uyIAVpIIOZZ5JqeqvYYIcEL727kskC66kF92t6Xl2V" crossorigin="anonymous"></script>
<script src="https://code.jquery.com/jquery-3.7.1.min.js" type="text/javascript"></script>
<script src="https://unpkg.com/axios/dist/axios.min.js"></script>
<script type="text/javascript">

    function myfunone(){
        return "Hello Sir, i am myfunone";
    }

    console.log(myfunone());

    // => async

    async function myfuntwo(){
        return "hello sir, i am my funtwo";
    }

    console.log(myfuntwo());  // Promise {<fulfilled>: 'hello sir, i am my funtwo'}

    myfuntwo()
        .then((response)=>console.log(response));


    async function myfunthree(){
        const setpromise = new Promise((resolve,reject)=>{
            setTimeout(()=>resolve("hello world"),3000);
        });

        const result = await setpromise;
        return result;
    }

    myfunthree()
        .then((response)=>console.log(response));

    async function myfunfour(){
        const setpromise = new Promise((resolve,reject)=>{
            setTimeout(()=>resolve('hello Earth'),3000);
        });

        const error = true;     // false

        if(!error){
            const response = await setpromise;
            return response;
        }else{
            await Promise.reject(new Error("something went Worng!!!"));
        }

    }

    // Promise {<pending>}
    // console.log(myfunfour()); // not working no result  // Promise {<pending>}

    myfunfour()
        .then(response=>console.log(response))
        .catch(err=>console.log(err));

</script>
</body>
</html>